cpp/mrc/src/internal/service.cpp

```diff
@@ -17,131 +17,295 @@
 
 #include "internal/service.hpp"
 
+#include "mrc/core/utils.hpp"
+#include "mrc/exceptions/runtime_error.hpp"
+#include "mrc/utils/string_utils.hpp"
+
 #include <glog/logging.h>
 
+#include <future>
+#include <mutex>
 #include <ostream>
 #include <utility>
 
 namespace mrc {
 
+Service::Service(std::string service_name) : m_service_name(std::move(service_name)) {}
+
 Service::~Service()
 {
+    if (!m_call_in_destructor_called)
+    {
+        LOG(ERROR) << "Must call Service::call_in_destructor to ensure service is cleaned up before being "
+                      "destroyed";
+    }
+
     auto state = this->state();
     CHECK(state == ServiceState::Initialized || state == ServiceState::Completed);
 }
 
+const std::string& Service::service_name() const
+{
+    return m_service_name;
+}
+
+bool Service::is_service_startable() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return (m_state == ServiceState::Initialized);
+}
+
+bool Service::is_running() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return (m_state > ServiceState::Initialized && m_state < ServiceState::Completed);
+}
+
+const ServiceState& Service::state() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return m_state;
+}
+
 void Service::service_start()
 {
-    if (forward_state(ServiceState::Running))
+    std::unique_lock<decltype(m_mutex)> lock(m_mutex);
+
+    if (!this->is_service_startable())
     {
-        do_service_start();
+        throw exceptions::MrcRuntimeError(MRC_CONCAT_STR(this->debug_prefix() << " Service has already been started"));
+    }
+
+    if (forward_state(ServiceState::Starting))
+    {
+        // Unlock the mutex before calling start to avoid a deadlock
+        lock.unlock();
+
+        try
+        {
+            this->do_service_start();
+
+            // Use ensure_state here in case the service itself called stop or kill
+            this->ensure_state(ServiceState::Running);
```

Comment:
Question: It sounds like `do_service_start` will advance the state, possibly to a `Stopping`, `Killing`, or `Completed` state which would surpass the `Running` state desired here, making this a no-op. Did you mean instead to assert the state?

Reply:
No, I added the new initializing state to help determine debugging issues that occur during process startup (i.e. if you throw an exception in `do_service_start`. So this just ensures that after that function returns successfully, we are set to running. If it doesnt return successfully, its possible we are already in a stopping or killing state which would mean asserting we move forward would just throw an error when there already is an error.